File: client/src/types/tracking.ts
export type ContainerStatus =
  | "moving"
  | "delayed"
  | "cleared"
  | "high_risk"
  | "idle";

export interface ContainerLive {
  id: string;
  container_code: string;
  status: ContainerStatus;
  cargo_type: string;
  lat: number;
  lng: number;
  speed: number;
  heading: number;
  origin_port: string | null;
  destination_port: string | null;
  ship_name: string | null;
  eta: string | null;
  distance_remaining_km: number | null;
  updated_at: string;
}

export interface TrackingSnapshot {
  type: "tracking_snapshot";
  server_time: string;
  containers: ContainerLive[];
}

export interface RouteWaypoint {
  lat: number;
  lng: number;
  timestamp: string;
}

export interface RouteHistory {
  container_id: string;
  container_code: string;
  waypoints: RouteWaypoint[];
  eta_predicted: string | null;
  delay_probability: number;
}

export const STATUS_LABELS: Record<ContainerStatus, string> = {
  moving: "Moving",
  delayed: "Delayed",
  cleared: "Cleared",
  high_risk: "High Risk",
  idle: "Idle",
};

File: server/app/schemas/tracking.py
from datetime import datetime

from pydantic import BaseModel


class ContainerLiveOut(BaseModel):
    id: str
    container_code: str
    status: str
    cargo_type: str
    lat: float
    lng: float
    speed: float
    heading: float
    origin_port: str | None = None
    destination_port: str | None = None
    ship_name: str | None = None
    eta: datetime | None = None
    distance_remaining_km: float | None = None
    updated_at: datetime


class TrackingSnapshot(BaseModel):
    type: str = "tracking_snapshot"
    server_time: datetime
    containers: list[ContainerLiveOut]


class RouteWaypoint(BaseModel):
    lat: float
    lng: float
    timestamp: datetime


class RouteHistoryOut(BaseModel):
    container_id: str
    container_code: str
    waypoints: list[RouteWaypoint]
    eta_predicted: datetime | None = None
    delay_probability: float
